File: packages/core-bridge/src/graph.ts
import type { CoreBridge } from "./types";

/** A node in the object graph — a slim projection, never the entity body (PLAN §5.2). */
export interface GraphNode {
  id: string;
  type: "note" | "task" | "habit" | "project";
  title: string;
  objectTypeId?: string | null;
  status?: string | null;
  /** Client-side display enrichment (not from core): the archetype's chosen color / icon,
   *  resolved from its object type so the renderer can mark objects distinctly. */
  objectColor?: string | null;
  objectIcon?: string | null;
}

/** A derived edge in the link index. Beyond the fixed kinds, reference-typed object props
 *  produce dynamic `prop:<field>` kinds (PLAN §4.1, §6.3). */
export type GraphEdgeKind = "ref" | "embed" | "stack" | "member" | `prop:${string}`;

/** A derived edge in the link index. */
export interface GraphEdge {
  sourceType: string;
  sourceId: string;
  targetType: string;
  targetId: string;
  kind: GraphEdgeKind;
}

export interface Graph {
  nodes: GraphNode[];
  edges: GraphEdge[];
}

/** Typed wrappers over the graph.* core methods (PLAN §5.2). */
export function graphApi(core: CoreBridge) {
  return {
    /** The whole knowledgebase graph (ids/titles/edges only). */
    full: () => core.invoke<Graph>("graph.full"),
    /** The subgraph within `depth` hops of a seed node. */
    neighborhood: (type: string, id: string, depth = 2) =>
      core.invoke<Graph>("graph.neighborhood", { type, id, depth }),
    /** Nodes that reference the given target ("linked mentions"). */
    backlinks: (type: string, id: string) =>
      core.invoke<GraphNode[]>("graph.backlinks", { type, id }),
    /** Title search — powers wikilink autocomplete. `type` scopes to one entity type
     * (omit or "all" for every type). */
    search: (query: string, type?: string, limit = 20) =>
      core.invoke<GraphNode[]>("graph.search", { query, type, limit }),
    /** Resolve a single node by id (any type); null when nothing matches. */
    lookup: (id: string) => core.invoke<GraphNode | null>("graph.lookup", { id }),
    /** Truncate and re-derive the index; returns the resulting counts. */
    rebuild: () => core.invoke<{ nodes: number; edges: number }>("graph.rebuild"),
  };
}

export type GraphApi = ReturnType<typeof graphApi>;

File: core/bridge/graph.go
package bridge

import "encoding/json"

// Graph read methods (PLAN §5.2). They project only the link index + graph_nodes, never
// entity bodies, so even graph.full stays small for the whole knowledgebase.

func (c *Core) graphFull() ([]byte, error) {
	g, err := c.store.Links.Full()
	if err != nil {
		return nil, err
	}
	return json.Marshal(g)
}

func (c *Core) graphNeighborhood(payload []byte) ([]byte, error) {
	var args struct {
		Type  string `json:"type"`
		ID    string `json:"id"`
		Depth int    `json:"depth"`
	}
	if err := unmarshal(payload, &args); err != nil {
		return nil, err
	}
	if args.Depth == 0 {
		args.Depth = 2
	}
	g, err := c.store.Links.Neighborhood(args.Type, args.ID, args.Depth)
	if err != nil {
		return nil, err
	}
	return json.Marshal(g)
}

func (c *Core) graphBacklinks(payload []byte) ([]byte, error) {
	var args struct {
		Type string `json:"type"`
		ID   string `json:"id"`
	}
	if err := unmarshal(payload, &args); err != nil {
		return nil, err
	}
	nodes, err := c.store.Links.Backlinks(args.Type, args.ID)
	if err != nil {
		return nil, err
	}
	return json.Marshal(nodes)
}

// graphSearch returns nodes whose title matches the query, across every entity type.
// Powers the editor's wikilink autocomplete (PLAN §5.2).
func (c *Core) graphSearch(payload []byte) ([]byte, error) {
	var args struct {
		Query string `json:"query"`
		Type  string `json:"type"`
		Limit int    `json:"limit"`
	}
	if err := unmarshal(payload, &args); err != nil {
		return nil, err
	}
	nodes, err := c.store.Links.Search(args.Query, args.Type, args.Limit)
	if err != nil {
		return nil, err
	}
	return json.Marshal(nodes)
}

// graphLookup resolves a single node by id (any type), for turning a pasted UUID into a
// typed wikilink. Returns JSON null when nothing matches.
func (c *Core) graphLookup(payload []byte) ([]byte, error) {
	var args struct {
		ID string `json:"id"`
	}
	if err := unmarshal(payload, &args); err != nil {
		return nil, err
	}
	node, err := c.store.Links.LookupNode(args.ID)
	if err != nil {
		return nil, err
	}
	return json.Marshal(node)
}

// graphRebuild truncates and re-derives the whole index (after migrations/imports or
// to recover from an extractor change). It signals a bulk data change so open graph
// views refresh.
func (c *Core) graphRebuild() ([]byte, error) {
	nodes, edges, err := c.store.Links.Rebuild()
	if err != nil {
		return nil, err
	}
	c.emitDataChanged("", "")
	return json.Marshal(map[string]int{"nodes": nodes, "edges": edges})
}
